airbyte-lib/airbyte_lib/_processors.py

```diff
@@ -0,0 +1,309 @@
+# Copyright (c) 2023 Airbyte, Inc., all rights reserved.
+
+"""Define abstract base class for Processors, including Caches and File writers.
+
+Processors can all take input from STDIN or a stream of Airbyte messages.
+
+Caches will pass their input to the File Writer. They share a common base class so certain
+abstractions like "write" and "finalize" can be handled in either layer, or both.
+"""
+
+from __future__ import annotations
+
+import abc
+import contextlib
+import io
+import sys
+from collections import defaultdict
+from typing import TYPE_CHECKING, Any, cast, final
+
+import pyarrow as pa
+import ulid
+
+from airbyte_protocol.models import (
+    AirbyteMessage,
+    AirbyteRecordMessage,
+    AirbyteStateMessage,
+    AirbyteStateType,
+    AirbyteStreamState,
+    ConfiguredAirbyteCatalog,
+    Type,
+)
+
+from airbyte_lib._util import protocol_util  # Internal utility functions
+
+
+if TYPE_CHECKING:
+    from collections.abc import Generator, Iterable, Iterator
+
+    from airbyte_lib.config import CacheConfigBase
+
+
+DEFAULT_BATCH_SIZE = 10000
+
+
+class BatchHandle:
+    pass
+
+
+class AirbyteMessageParsingError(Exception):
+    """Raised when an Airbyte message is invalid or cannot be parsed."""
+
+
+class RecordProcessor(abc.ABC):
```

Comment:
Starting a new thread so it shows up properly in the GH UI (related to https://github.com/airbytehq/airbyte/pull/33607#discussion_r1447800745 )

Thanks for explaining a bit, I think I get it now - the weird part is the fact that both the Cache instance and the FileWriter instance are RecordProcessors, but the cache is managing the file writer. IMHO this makes the logic very difficult to follow because it's the same methods but they are called on different objects managing different state.

I think this comes from the fact that we wanted to provide file caches at some point as well, but it strikes me as too "clever" - even though it prevents redundancy, it makes it hard to follow to control flow.

We don't need to refactor on this PR because like you I would really like to get that one in, but IMHO we should make sure that the bug is fixed before we release.

I tried to make sense of the control flow once more and this is what happens:
* A batch is processed on the cache instance (`_process_batch`)
* It is delegating the actual writing of the batch to the underlying file writer (`_write_batch`), but returns the batch handle returned from that, so now the `_pending_batches` of the cache instance are holding the file writer batch handles pointing to the files
* Afterwards it's finalizing the batches by writing them to the actual duckdb (`_finalize_batches`), calling `_finalizing_batches` for cleanup, which does call `_cleanup_batch` with the file writer batch handles, but it calls it on the cache instance, not on the file writer instance, so it just executes the noop and leaves the files alone

You wrote this:
> The file writer does have a cleanup method as of now, and it inherits from here. So when this runs in the file writer class, it should (in theory) correctly delete the files. I didn't have time to test but I at least wanted to get the implementation "hooks" in place.

The issue is that `_cleanup_batch` on the file writer is never called with the batch handle, as the cache manages it and it doesn't delegate the call.

The easiest way to fix it now is what I wrote in the other comment - just like `_write_batch` on the cache delegates to the underlying file writer, `_cleanup_batch` needs to as well:
```python
    @overrides
    def _write_batch(
        self,
        stream_name: str,
        batch_id: str,
        record_batch: pa.Table | pa.RecordBatch,
    ) -> FileWriterBatchHandle:
        """
        Process a record batch.

        Return the path to the cache file.
        """
        return self.file_writer.write_batch(stream_name, batch_id, record_batch)

    # this is the new part
    def _cleanup_batch(
        self,
        stream_name: str,
        batch_id: str,
        batch_handle: BatchHandle,
    ) -> None:
        self.file_writer._cleanup_batch(stream_name, batch_id, batch_handle)
```

because the file writer is not "remembering" the patches it worked on, it's just getting a `_write_batch` call and processes it, so the cache needs to make sure it's called.

The cache might do more cleanup stuff per batch at some point, but it needs to make sure it's also done for the file write instance it manages.

As mentioned above, I think we should revisit this architecture, but doesn't have to happen on this PR as long as we fix the bug.

Reply:
Regarding the design, I'm open to other methods but I think there are some benefits to this model.

I'll fix the plumbing for the batch handing.

What I didn't have time for yesterday, but I can do today, is just to add an integration test that asserts files are deleted as expected.

Reply:
Sounds good, +1 for an integration test

Reply:
Done. 👍 

Tests added and passing.